src/client/game.cpp

```diff
@@ -1381,6 +1381,23 @@ bool Game::createSingleplayerServer(const std::string &map_dir,
 			false, nullptr, error_message);
 	server->start();
 
+	input->clear();
+
+	FpsControl fps_control;
+	f32 dtime;
+	fps_control.reset();
+
+	while (!server->isInitialized()) {
+		if (!m_rendering_engine->run())
+			return false;
+		if (input->cancelPressed())
+			return false;
+
+		fps_control.limit(device, &dtime);
+		showOverlayMessage(N_("Creating server..."), dtime, 5);
+	}
+
+	// TODO: This is IO on the main thread, can we move it to the server thread?
```

Comment:
To not change the server init too much and also to incorporate this I'd move server init into a short lived thread here

Reply:
Hmm, I guess you mean something like https://github.com/minetest/minetest/commit/dfc3f58e2166004b2ad70253b513db9b08558ee4 ?

It's certainly easier to integrate `copyServerClientCache()` this way. However, it means that I have to handle "early exit" and the shutdown screen manually, which will probably cause problems if we ever want to fix shutdown blocking the main thread. By using the server thread, I get "early exit" and the shutdown screen for free.

Reply:
Yes, like that. For ergonomics I'd probably try to write a helper that turns a lambda into a startable thread,

> However, it means that I have to handle "early exit" and the shutdown screen manually, which will probably cause problems if we ever want to fix shutdown blocking the main thread.

Are you sure?
No matter where server startup runs we can't interrupt it and have to wait for it before stopping the server.

Or did you mean something else?

Reply:
> No matter where server startup runs we can't interrupt it and have to wait for it before stopping the server.

Yes, that's true. What I mean is that if we do server startup on the server thread, we don't need extra code to wait for startup to finish before shutting down. It Just Works because `Server::~Server` already tells the server thread to stop.

With the "short-lived thread" approach you suggest, I don't see a clean way of waiting for startup to finish before shutting down. I have implemented a way in dfc3f58e2166004b2ad70253b513db9b08558ee4, but I don't like it.

Reply:
What do you think about my new commit?

Reply:
:+1: